packages/types/Polymorphism.ts

```diff
@@ -0,0 +1,165 @@
+import type {
+  ComponentPropsWithoutRef,
+  ElementType,
+  Fragment,
+  ReactElement,
+  ReactPortal,
+} from 'react'
+import type { Maybe, Overwrite, RefOf } from '.'
+
+// ----- Internal Utilities -----
+
+/**
+ * Adjustment to `ref` property on `Fragment` to allow developers to forward the ref
+ * on the root element without needing to use conditional logic.
+ *
+ * @internal
+ */
+type FragmentRef<TElement extends ElementType> = TElement extends typeof Fragment
+  ? { ref?: never }
+  : {}
+
+/**
+ * Applies signature polymorphism to a component. If the `as` property is provided,
+ * the component will render as the provided element type. If the `as` property is
+ * not provided, the component will render as the default element type.
+ *
+ * @internal
+ */
+type Tag<TActualElement, TInternalElement extends ElementType> = TActualElement extends ElementType
+  ? TActualElement
+  : TInternalElement
+
+/**
+ * A polymorphic component's JSX-equivalent return type.
+ * Functionally similar to `ReactElement`.
+ */
+export type PolymorphicElement<TElement extends ElementType, TProps extends object> =
+  | ReactElement<TProps, TElement>
+  | ReactPortal
+  | null
+  | undefined
+
+// ----- Simple Polymorph -----
+
+/**
+ * A polymorphic component type definition which adds an `as` property to the
+ * component definition.
+ *
+ * @example
+ * ```tsx
+ * type Props = {
+ *   greeting: string;
+ * }
+ *
+ * const Greeting: PolymorphicComponent<'h1', Props> = (props) => {
+ *   const {
+ *     as: Tag = 'h1',
+ *     greeting,
+ *     ...rest
+ *   } = props
+ *
+ *   return <Tag {...rest}>{greeting}</h1>;
+ * };
+ *
+ * const App = () => {
+ *   return <Greeting as="h2" greeting="Hello, world!" />
+ * };
+ * ```
+ */
+export type PolymorphicComponent<TInternalElement extends ElementType, TProps extends object> = <
+  TActualElement = TInternalElement,
+>(
+  props: Overwrite<
+    ComponentPropsWithoutRef<Tag<TActualElement, TInternalElement>> & {
+      as?: Tag<TActualElement, TInternalElement>
+    },
+    TProps
+  >,
+) => PolymorphicElement<Tag<TActualElement, TInternalElement>, TProps>
+
+// ----- Forward Polymorph -----
```

Comment:
**thought:** Could this be a bit misleading? 

We aren't really forwarding a polymorph. We're forwarding _a ref_ to a polymorph.

Reply:
As a first version we're publishing without the poly helper.